### ISSUE TITLE
networking problem  .net core with Centos os

### ISSUE THREAD

**Issue by FinexTech** [dotnet/corefx#41207]
# Issue Title

Absolutely low performance on Centos operation systems

# General

We are using .net core & Centos operation system.
We are doing load testing. We have bots, bots are developed on .net core 2.1 & hosted on cetnos os. Bots open connection to another application on another machine and each bot has opened websocket connection to another application. We try to emulate real users. Bots send rest request to another application and receive results via web socket. 
When we run our load tool on centos 200 bots took 100% machine utilization.
When we run the same load tool on windows os 1000 bots took 40% machine utilization.
I suggest that it is problem with linux network stack & . net core.
May be we need to do some tuning of our linux machine. 
Or you know what could be causing this?
Thanks, Ivan.

**Comment by scalablecory** [comment]
Can you try with .NET Core 3.0? You may have better performance there.

**Comment by FinexTech** [comment]
Thanks, I am planning to rerun tests today on version 3.0. Will let you know ASAP.

**Comment by wfurt** [comment]
Also we would need some more specific. This is not actionable as written. It would be great if you can isolate it to some simple repro so it is at least clear what exactly do you use and how.
If you do lot of parallel processing you may watch for ThreadPool size.

**Comment by FinexTech** [comment]
Here are some observations:
- We updated our code base to use netcore 3.0. This step improved performance twice but we still experienced load spikes.
- We changed linux (centos) kernel from 3.x to 5.x. This change removed all spikes and machine utilization became smooth. It improved performance drastically.
- We did some changes to linux network configuration: extended local port range, decreased fin timeout, allowed port reuse for outgoing connections, hardcoded hosts into /etc/hosts file. These fixed an issue with ports / request timeouts / name resolution (dns name resolution worked for the first minute only)

After all these changes application performance became pretty close to what we have on windows machine.

Thank you.

**Comment by wfurt** [comment]
do you have idea what specifically changed with kernel upgrade @FinexTech ?  
It would be good to know for future reference. 

cc: @tmds

**Comment by FinexTech** [comment]
Right now I can't say what changed with kernel upgrade. We are planning to do more tests in 1-2 weeks. Will try to use oracle linux, ubuntu and play a little bit more with various kernel versions. I will be back when there are any results.

**Comment by tmds** [comment]
@FinexTech fyi, CentOS 8 was just released.

**Comment by FinexTech** [comment]
> @FinexTech fyi, CentOS 8 was just released.

Thank you, @tmds. Will try it too.

**Comment by scalablecory** [comment]
@FinexTech were you able to do any testing?

**Comment by FinexTech** [comment]
We dumped all sysctl parameters with 2 different versions of kernel (v3 and v5) for comparison. The main idea was that some defaults were changed in new version. An attempt to apply new values didn't give any results. Kernel v4 works the same way as v3.
All these performance spikes are somehow related to outgoing connections (we use restsharp for doing API requests).
BTW: every rest request does domain name resolution, and on linux machine in 1 minute dns server stops giving us a reply. Do we need to use some local DNS caching? Windows machine doesn't have such a problem. I thought dotnet uses connection pool to store and reuse connections.

**Comment by wfurt** [comment]
Yes, it seems like in this case DNS would be useful. .NET fully depends on OS and it does not cache any answers.  In norma cases persistent connection would be maintain and lookup for each request would be avoided but it seems like this is not quite same for you. The pool would be maintained but it depends on how you structure you code and if you use same HttpClient for subsequent requests. 
Also on Ubuntu the default resolver points to DNS cache from systemd but I think Centos does not have it by default. I'm sure you can use Bind in caching mode as well as there are other implementations.
